go/vt/vtgate/evalengine/expr_collate.go

```diff
@@ -44,7 +44,7 @@ var collationBinary = collations.TypedCollation{
 
 var collationJSON = collations.TypedCollation{
 	Collation:    46, // utf8mb4_bin
-	Coercibility: collations.CoerceCoercible,
+	Coercibility: collations.CoerceImplicit,
```

Comment:
Turns out that in practice JSON uses a different coercibility. Changing this fixes a whole slew of issues otherwise exposed by the `CONCAT` tests with `JSON` fields.